src/TopDownProteomics/ProForma/ChemicalProteoformHashGenerator.cs

```diff
@@ -0,0 +1,82 @@
+                this.GetFormulaDescriptor(proteoformGroup.NTerminalModification);
+            IList<ProFormaDescriptor> nTermDescriptors = nTermDescriptor == null ? null : new[] { nTermDescriptor };
+
+            ProFormaDescriptor cTermDescriptor =
+                this.GetFormulaDescriptor(proteoformGroup.CTerminalModification);
+            IList<ProFormaDescriptor> cTermDescriptors = cTermDescriptor == null ? null : new[] { cTermDescriptor };
+
+            IList<ProFormaTag> tags = new List<ProFormaTag>();
+            if (proteoformGroup.Modifications != null)
+            {
+                foreach (IProteoformModificationWithIndex proteoformModificationWithIndex in proteoformGroup.Modifications)
+                {
+                    ProFormaDescriptor descriptor = this.GetFormulaDescriptor(proteoformModificationWithIndex);
+                    tags.Add(new ProFormaTag(proteoformModificationWithIndex.ZeroBasedIndex, new[] { descriptor }));
+                }
+            }
+
+            ProFormaTerm proFormaTerm = new ProFormaTerm(proteoformGroup.GetSequence(), null, nTermDescriptors, cTermDescriptors, tags.OrderBy(t => t.Descriptors.First().Value).ToArray());
+            string hash = new ProFormaWriter().WriteString(proFormaTerm);
+            return hash;
+        }
+
+        private ProFormaDescriptor GetFormulaDescriptor(IProteoformModification proteoformModification)
+        {
+            return proteoformModification == null
+                    ? null
+                    : new ProFormaDescriptor(ProFormaKey.Formula, this.GetChemicalFormulaString(proteoformModification.GetChemicalFormula()));
+        }
+
+        private string GetChemicalFormulaString(IChemicalFormula chemicalFormula)
+        {
+            ICollection<IEntityCardinality<IElement>> elements = chemicalFormula.GetElements().ToList();
+            StringBuilder formula = new StringBuilder();
+            IEntityCardinality<IElement> carbon = elements.SingleOrDefault(e => e.Entity.AtomicNumber == 6);
+
+            if (carbon != null && carbon.Count > 0)
+            {
+                this.AppendFormula(formula, carbon);
+                elements.Remove(carbon);
+                IEntityCardinality<IElement> hydrogen = elements.SingleOrDefault(e => e.Entity.AtomicNumber == 1);
+                if (hydrogen != null && hydrogen.Count > 0)
+                {
+                    this.AppendFormula(formula, hydrogen);
+                    elements.Remove(hydrogen);
+                }
+            }
+            
+            foreach (IEntityCardinality<IElement> element in elements.OrderBy(e => e.Entity.Symbol))
+            {
+                if (element.Count > 0)
+                {
+                    this.AppendFormula(formula, element);
+                }
+            }
+
+            return formula.ToString();
+        }
+
+        private void AppendFormula(StringBuilder formula, IEntityCardinality<IElement> element)
```

Comment:
Consider creating this as a local function inside of `GetChemicalFormulaString()`
[Local Function Docs](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions)